New Deposits Overview UI Foundation Component

mattallan opened the pull request:
Fixes #5694 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR introduces the base file structure and initial component for the new deposits overview UI.

The file structure I decided to go with was to have separate directories for each sub-component of the new Deposits overview UI:

 - `components/deposits-overview` is the main directory for the component
   - `components/deposits-overview/next-deposit/`
   - `components/deposits-overview/deposit-history/`
   - `components/deposits-overview/actions/`

So far this is what the skeleton looks like:

![image](https://user-images.githubusercontent.com/2275145/224861044-ff498b26-ef1f-4bda-b44e-70578777e38e.png)

But each component/directory is just a skeleton for now and maps to the final designs as follows:

![image](https://user-images.githubusercontent.com/2275145/224668761-ed60512d-9109-44a5-8f42-03bccc430cc1.png)


#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. `npm run build:client`
2. Set `_wcpay_feature_simplify_deposits_ui` in the `wp_options` table to `1`
3. Load the Payments Overview page

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
Add dummy feature flag

Commit:
Add skeleton structure/components for the new deposits overview UI changes

Commit:
Load the new deposits overview components if the feature flag is enabled

Commit:
default feature flag to false

Commit:
Merge branch 'develop' into issue/5694-new-deposits-card-skeleton

Commit:
Empty changelog

haszari commented on changelog/issue-5694-new-deposits-card-skeleton:
```diff
@@ -0,0 +1,5 @@
+Significance: patch
+Type: dev
+Comment: This PR is adding low level structural changes to implement deposits overview changes
```
Does this comment go out in changelog, or does `patch` mean it's hidden? Could add the eventual changelog here, e.g. `New deposits overview card (behind feature flag)`

mattallan replied:
No, this comment shouldn't get written to any changelogs.

When filling out this changelog, I chose an empty changelog message and it said that the changelog won't be written but it wanted to know a reason.

haszari commented on client/overview/index.js:
```diff
@@ -131,7 +132,10 @@ const OverviewPage = () => {
 			{ ! accountRejected && (
 				<ErrorBoundary>
-					<DepositsInformation />
+					{ window.wcpaySettings?.featureFlags?.isNewDepositsOverview
```
Nice!

haszari commented on includes/class-wc-payments-features.php:
```diff
@@ -254,6 +254,7 @@ public static function to_array() {
 				'woopayExpressCheckout'   => self::is_woopay_express_checkout_enabled(),
 				'isAuthAndCaptureEnabled' => self::is_auth_and_capture_enabled(),
 				'progressiveOnboarding'   => self::is_progressive_onboarding_enabled(),
+				'isNewDepositsOverview'   => false,
```
Are you planning on including the feature flag in this PR (or is this a placeholder)?

mattallan replied:
Nah, I had a note in the PR description but it could've been missed.

The real feature flag is being implemented by https://github.com/Automattic/woocommerce-payments/issues/5690

mattallan replied:
Shendy has pushed up a PR for the feature flag so I've updated this PR to use the correct flag. See https://github.com/Automattic/woocommerce-payments/pull/5763 for implementation

haszari reviewed (commented):
Added some comments. 

- Ideally can leave the child components up to future PRs (naming, files, composition etc). 
- IMO getting the feature flag and outer card in place is the key goal for this issue/PR.

haszari commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,84 @@
+import NextDepositDetails from './next-deposit';
+import DepositsHistory from './history';
+import DepositsOverviewActions from './actions';
+import { useAllDepositsOverviews } from 'wcpay/data';
```
I'd suggest just hard-coding placeholder content in this component, and let future PRs flesh that out as components. 

By adding placeholder components (and naming them) you're constraining future PRs - it's like a specification.

mattallan replied:
Removed in [1ec0e0a](https://github.com/Automattic/woocommerce-payments/pull/5759/commits/1ec0e0a881a1cece7e2e8505e4b75806715fdc91)

Commit:
Remove stub child components from initial Deposits Overview PR

Commit:
Fix JS linting issues

Commit:
Merge branch 'develop' into issue/5694-new-deposits-card-skeleton

Commit:
Update featureflag usage as implemented by #5763

Commit:
Remove dropdown menu as they don't exist in latest designs

mattallan commented:
I was looking at outdated designs for this card and there are no vertical ellipses in the card header. I've removed these in [2612b0d](https://github.com/Automattic/woocommerce-payments/pull/5759/commits/2612b0d9ac31b4f217444dd3c1ea142baac81b66)

I've also updated the dummy featureflag now that https://github.com/Automattic/woocommerce-payments/pull/5763 has been implemented.

Commit:
Use simplifyDepositsUi from defined wcpaySettings variable

Commit:
Merge branch 'develop' into issue/5694-new-deposits-card-skeleton

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,65 @@
+	const { currencies, account } = overviews;
+
+	return (
+		<React.Fragment>
```
Nit: I'd suggest replacing `<React.Fragment>` with `<>`, as it seems to be more common in this repo.
```suggestion
		<>
```

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,65 @@
+
+	return (
+		<React.Fragment>
+			{ currencies.map( ( overview: AccountOverview.Overview ) => (
```
This will render a `<DepositsOverviewDetails>` card for each currency, rather than just one for the "selected" currency.

This is probably not a problem for this skeleton component issue and can be addressed later in #5751.

mattallan replied:
Thanks for picking this up @Jinksi! For now, I've updated this to just get the first current from the list instead of looping over the array.

https://github.com/Automattic/woocommerce-payments/pull/5759/commits/6c577b866472d1da49db12ca8db262503d6ceace

We will have to figure out how we'll get the current currency that has been selected on the Account Balances, but that can come later I think.

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,65 @@
+	return (
+		<Card>
+			<CardHeader>
+				{ __( 'Deposits', 'woocommerce-payments' ) }
```
This is a good candidate for a `strings.ts` file within the component dir.

Or we can create this file in upcoming issues if needed.

mattallan replied:
Thanks for the suggestion. I've added a start to this in [b8aaf2a](https://github.com/Automattic/woocommerce-payments/pull/5759/commits/b8aaf2a112178bcb24f2c286cf2ecee222835591)

My thinking with the structure of this `strings.ts` file would be something like:

```
export default {
	heading: __( 'Deposits', 'woocommerce-payments' ),
	next_deposits: {
		heading: __( 'Next deposits', 'woocommerce-payments' ),
		description: __( 'The amount may change while payments are still accumulating', 'woocommerce-payments' ),
	},
	deposits_history: {
		heading: __( 'Deposits history', 'woocommerce-payments' ),
	},
};
```

Commit:
Show the deposits overview card in a loading state

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Renders a deposits overview
+ *
+ * @param {AccountOverview.Overview} props Deposits overview
```
Suggestion: I think this JSDoc param type should match the type provided to the function declaration below: `const DepositsOverviewDetails: React.FunctionComponent< OverviewProps > = (`
```suggestion
 * @param {OverviewProps} props Deposits overview
```

mattallan replied:
Fixed in [8459303](https://github.com/Automattic/woocommerce-payments/pull/5759/commits/8459303f34be76f2e563134f32dac3e09a8e3a7b)

Commit:
Update Deposits Overview foundation from feedback

Commit:
Move strings to a new strings.ts file

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,73 @@
+			/>
+		);
+	}
+
```
Nit: I know this is a temporary solution, but I'd suggest avoiding [prop/state mutation in React as a general rule of thumb](https://reactjs.org/docs/components-and-props.html#props-are-read-only).
```suggestion
	const overview = currencies[ 0 ]; // TODO: To handle multiple currencies we'll need to fetch the currently selected currency.
```

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,73 @@
+		);
+	}
+
+	const overview = currencies.shift(); // TODO: To handle multiple currencies we'll need to fetch the currently selected currency.
```
Nit: I know this is a temporary solution, but I'd suggest avoiding [prop/state mutation in React as a general rule of thumb](https://reactjs.org/docs/components-and-props.html#props-are-read-only).
```suggestion
	const overview = currencies[ 0 ]; // TODO: To handle multiple currencies we'll need to fetch the currently selected currency.
```

haszari replied:
I think that's more than a rule of thumb, very dangerous! It's a hard rule IMO, good catch 🙌

Commit:
Update client/components/deposits-overview/index.tsx

Co-authored-by: Eric Jinks <3147296+Jinksi@users.noreply.github.com>

Commit:
Fix type in JSdoc param

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,73 @@
+ */
+import * as React from 'react';
+import { Card, CardHeader } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
```
Nit: this import is now unused since moving the strings to `strings.ts`.

```suggestion

```

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -0,0 +1,73 @@
+
+	if ( isLoading ) {
+		return (
+			<DepositsOverviewDetails
```
A note for the future:

When we have the UI finalised, we should create a separate component that is rendered only while `isLoading` is true, using the `<Loadable />` component for the "loading" animation UI.

For example: https://github.com/Automattic/woocommerce-payments/blob/8459303f34be76f2e563134f32dac3e09a8e3a7b/client/components/deposits-information/loading.tsx

mattallan replied:
I wasn't actually sure if we'd need to do something different here because the designs for the isLoading state only include `<Loadable />` components for the Next Deposits heading, description, and also the Next deposits table (screenshot below), so I was thinking the Next Deposits child component would have a loading component since it will have all of the styling needed for the table etc.

![image](https://user-images.githubusercontent.com/2275145/224897842-59239827-8608-46db-a9ca-bf0e55ac6c13.png)

Jinksi replied:
Ah yeh, I see, thanks! Pass the isLoading state to those components to decide 💯

mattallan replied:
Yeah that's what I was thinking but I'm not actually sure which is the correct approach for this sort of thing. If you're okay with it then I guess we can go with this and iterate on it when we get to it

Commit:
Update client/components/deposits-overview/index.tsx

Co-authored-by: Eric Jinks <3147296+Jinksi@users.noreply.github.com>

Jinksi reviewed (approved):
LGTM @mattallan! 🏆 

I've tested with the `_wcpay_feature_simplify_deposits_ui` option values:
- [x] undefined: the existing `<DepositsInformation />` is rendered.
- [x] '0': the existing `<DepositsInformation />` is rendered.
- [x] '1': the new skeleton card is rendered.

mattallan commented:
Thanks @Jinksi for all of the feedback! Merging this one into `develop`
